File: packages/dprint-plugin-typescript/src/utils/stringUtils.ts
export function isStringEmptyOrWhiteSpace(text: string) {
    const hasNonWhiteSpaceChar = /\S/.test(text);
    return !hasNonWhiteSpaceChar;
}

export function hasNewlineOccurrencesInLeadingWhitespace(text: string, occurrences: number) {
    if (occurrences === 0)
        hasNoNewlinesInLeadingWhitespace(text);

    for (let i = 0; i < text.length; i++) {
        if (!isStringEmptyOrWhiteSpace(text[i]))
            return false;
        if (text[i] === "\n" && --occurrences === 0)
            return true;
    }

    return false;
}

export function hasNoNewlinesInLeadingWhitespace(text: string) {
    for (let i = 0; i < text.length; i++) {
        if (!isStringEmptyOrWhiteSpace(text[i]))
            return true;
        if (text[i] === "\n")
            return false;
    }

    return true;
}

export function hasNewLineOccurrencesInTrailingWhitespace(text: string, occurrences: number) {
    if (occurrences === 0)
        return hasNoNewlinesInTrailingWhitespace(text);

    for (let i = text.length - 1; i >= 0; i--) {
        if (!isStringEmptyOrWhiteSpace(text[i]))
            return false;
        if (text[i] === "\n" && --occurrences === 0)
            return true;
    }

    return false;
}

export function hasNoNewlinesInTrailingWhitespace(text: string) {
    for (let i = text.length - 1; i >= 0; i--) {
        if (!isStringEmptyOrWhiteSpace(text[i]))
            return true;
        if (text[i] === "\n")
            return false;
    }

    return true;
}

File: crates/dprint-plugin-typescript/src/utils/string_utils.rs
pub fn has_new_line_occurrences_in_leading_whitespace(text: &str, occurrences: i8) -> bool {
    if occurrences == 0 {
        return has_no_new_lines_in_leading_whitespace(text);
    }

    let mut found_occurrences = 0;
    for c in text.chars() {
        if !c.is_whitespace() {
            return false;
        }
        if c == '\n' {
            found_occurrences += 1;
            if found_occurrences >= occurrences {
                return true;
            }
        }
    }

    return false;
}

pub fn has_no_new_lines_in_leading_whitespace(text: &str) -> bool {
    for c in text.chars() {
        if !c.is_whitespace() {
            return true;
        }
        if c == '\n' {
            return false;
        }
    }

    return true;
}

pub fn has_new_line_occurrences_in_trailing_whitespace(text: &str, occurrences: i8) -> bool {
    if occurrences == 0 {
        return has_no_new_lines_in_trailing_whitespace(text);
    }

    let mut found_occurrences = 0;
    for c in text.chars().rev() {
        if !c.is_whitespace() {
            return false;
        }
        if c == '\n' {
            found_occurrences += 1;
            if found_occurrences >= occurrences {
                return true;
            }
        }
    }

    return false;
}

pub fn has_no_new_lines_in_trailing_whitespace(text: &str) -> bool {
    for c in text.chars().rev() {
        if !c.is_whitespace() {
            return true;
        }
        if c == '\n' {
            return false;
        }
    }

    return true;
}
